### DISCUSSION TITLE
How to get a gix::Tag from a Reference?

### DISCUSSION THREAD

**Discussion by ghost** [GitoxideLabs/gitoxide#1397] [Q&A] [answered]
I am trying to crawl the tags for a git repository and get information within gix::Tag. However when I try and get an object with something like:

```rust
// where reference is a gix::Reference with category gix_ref::Category::Tag
let id = gix_reference.id();
// I expect this object should also be a tag
let object = id.object().unwrap();
// this panics as the object is not a tag but is a commit
let tag = object.try_into_tag().unwrap();
```

How do I actually get a gix::Tag object?

**Comment by Byron** [comment]
It's hard to tell what's going on in the snippet as it's all based on what `gix_reference` ends up being.
Maybe it was peeled before so it now points to what the tag pointed to.
Alternatively it's a lightweight tag, which points to the commit without a tag object being present. Thus you'd want to match on the kind of `object` obtained through `id.object()` to know what to do.

**Comment by ghost** [comment]
Is there an example or could you point me to how I can crawl all tags in a git repo and get the tag info not the commit info?

This is what I am currently doing:

```rust
// open repo
let repo = gix::open(repo_path).unwrap();
// get an iter over references in this repo
let ref_iter = repo.references().unwrap();
// filter to just tags and crawl them
for tag_ref in refs.tags().unwrap() {
  // get the category of the reference
  if let Some(category) = tag_ref.name().category() {
    // filter to just tags which makes me think the object should be a tag?
    if category == gix_ref::Category::Tag {
      // get this references id
      let id = tag_ref.id();
      // get this references object which I want to be a Tag but its a Commit
      // I understand this also has kind but I can't seem to figure out how to get a Tag
      let object = id.object().unwrap();
    }
  }
}
```

**Reply by Byron** [reply, marked as answer]
[This explains the culprit](https://chatgpt.com/share/469dd5eb-74fd-4458-ab10-7493c5cc2711).

**Reply by ghost** [reply]
thanks I was under the impression that lightweight tags were still backed by full Tag objects and were simply a cheaper to retrieve representation of tags.
